cdc/capture/capture.go

```diff
@@ -0,0 +1,340 @@
+			}
+			return errors.Trace(err)
+		}
+		// Campaign to be an owner, it blocks until it becomes the owner
+		if err := c.campaign(ctx); err != nil {
+			switch errors.Cause(err) {
+			case context.Canceled:
+				return nil
+			case mvcc.ErrCompacted:
+				continue
+			}
+			log.Warn("campaign owner failed", zap.Error(err))
+			// if campaign owner failed, restart capture
+			return cerror.ErrCaptureSuicide.GenWithStackByArgs()
+		}
+
+		log.Info("campaign owner successfully", zap.String("capture-id", c.info.ID))
+		owner := c.newOwner()
+		c.setOwner(owner)
+		err = c.runEtcdWorker(ctx, owner, model.NewGlobalState(), ownerFlushInterval)
+		c.setOwner(nil)
+		log.Info("run owner exited", zap.Error(err))
+		// if owner exits, resign the owner key
+		if resignErr := c.resign(ctx); resignErr != nil {
+			// if regisn owner failed, return error to let capture exits
+			return errors.Annotatef(resignErr, "resign owner failed, capture: %s", c.info.ID)
+		}
+		if err != nil {
+			// for errors, return error and let capture exits or restart
+			return errors.Trace(err)
+		}
+		// if owner exits normally, continue the campaign loop and try to election owner again
+	}
+}
+
+func (c *Capture) runEtcdWorker(ctx cdcContext.Context, reactor orchestrator.Reactor, reactorState orchestrator.ReactorState, timerInterval time.Duration) error {
+	etcdWorker, err := orchestrator.NewEtcdWorker(ctx.GlobalVars().EtcdClient.Client, kv.EtcdKeyBase, reactor, reactorState)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if err := etcdWorker.Run(ctx, c.session, timerInterval); err != nil {
+		// We check ttl of lease instead of check `session.Done`, because
+		// `session.Done` is only notified when etcd client establish a
+		// new keepalive request, there could be a time window as long as
+		// 1/3 of session ttl that `session.Done` can't be triggered even
+		// the lease is already revoked.
+		switch {
+		case cerror.ErrEtcdSessionDone.Equal(err),
+			cerror.ErrLeaseExpired.Equal(err):
+			return cerror.ErrCaptureSuicide.GenWithStackByArgs()
+		}
+		lease, inErr := ctx.GlobalVars().EtcdClient.Client.TimeToLive(ctx, c.session.Lease())
+		if inErr != nil {
+			return cerror.WrapError(cerror.ErrPDEtcdAPIError, inErr)
+		}
+		if lease.TTL == int64(-1) {
+			log.Warn("session is disconnected", zap.Error(err))
+			return cerror.ErrCaptureSuicide.GenWithStackByArgs()
+		}
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (c *Capture) setOwner(owner *owner.Owner) {
+	c.ownerMu.Lock()
+	defer c.ownerMu.Unlock()
+	c.owner = owner
+}
+
+// OperateOwnerUnderLock operates the owner with lock
+func (c *Capture) OperateOwnerUnderLock(fn func(*owner.Owner) error) error {
+	c.ownerMu.Lock()
+	defer c.ownerMu.Unlock()
+	if c.owner == nil {
+		return cerror.ErrNotOwner.GenWithStackByArgs()
+	}
+	return fn(c.owner)
+}
+
+// Campaign to be an owner
+func (c *Capture) campaign(ctx cdcContext.Context) error {
+	failpoint.Inject("capture-campaign-compacted-error", func() {
```

Comment:
Can you add comment on why we should test handling `ErrCompacted`? Under what circumstances can `ErrCompacted` happen in reality?